### DISCUSSION TITLE
Get Webview2 Process ID?

### DISCUSSION THREAD

**Discussion by KiddoV** [wailsapp/wails#1511] [Q&A] [answered]
So Wails generate some ``webview2`` processes. I make an app that has a tracking feature to track for memory and cpu usage (in percentage, specifically).
Right now, I can get the process of my app easily with ``os.Getpid()``, and get those info using ``gopsutil``. The thing is, if Wails has another processes such as ``webview2`` associated with my app that running separately, the information that I get right now will not be accurate since ``webview2`` is from a different process.
The question is, how do I get the ``webview2`` pid associated with my Wails app, so I can get the exact total of memory and cpu usage?

**Comment by stffabi** [comment, marked as answer]
You need to incorporate the parent process id of the processes and use the process tree. All WebView2 processes have a parent process and will finally root into the Wails pid. This could be over several levels of processes, e.g. WebView2 -> WebView2 -> Wails.

AFAIK gopsutil returns the parent process id, so you should be able to loop over all processes and find those which are connected the the wails process.

**Reply by KiddoV** [reply]
How do you know if that ``webview2`` connected to a specific Wails app? 
What if I have 2 or more Wails apps open at the same time, can we identify which one is associated which app?

**Reply by stffabi** [reply]
The root process id won't be the same, it's either the one of the WailsApp1 or WailsApp2.

```
pid = process id
ppid = parent process id

WailsApp1 (pid=22), WebView2 (pid=55, ppid=22), WebView2 (pid=66, ppid=55)
WailsApp2 (pid=122), WebView2 (pid=155, ppid=122), WebView2 (pid=166, ppid=155)
```

**Reply by KiddoV** [reply]
Oh... I see it now, so we track the process by its parent process, and so on... Didn't know ``os.Getppid()`` is a thing. :)
Thank you!

**Reply by KiddoV** [reply]
@stffabi  there is still a problem with this method. 
All the ``msedgewebview2.exe``(s) are not associated with the Wails app at all, weird!, instead, it is associated with ``explorer.exe``.
For example, in a session, my app ID was 45296 (which is the my-app-dev.exe - my app name in dev) I can track down the parent of it by using this command in ``cmd.exe``: ``wmic process get ProcessID, ParentProcessID | find "45296"`` --> return parent id: 35764 (which is the wails.exe).
When I use that command again for that parent, it will print out: (Left side is the parent id of the right one)
```cmd
> wmic process get ProcessID, ParentProcessID | find "45296"
2608 <-powershell.exe             35764 <-wails.exe
35764<-wails.exe            45296 <-my-app-dev.exe 
```
...Noted the root ``2608`` is the pid of my ``powershell.exe``
In this case, I expected it will print out all the associated child pid, but it only print 1 pid which is the  ``my-app-dev.exe``.

... So I went straight to the ``Task Manager`` to see what is the pid(s) for all of my ``msedgewebview2.exe``(s) are. I found the parent of all which is 41428 (msedgewebview2.exe)
I used the command again, and I got:
```cmd
> wmic process get ProcessID, ParentProcessID | find "41428"
6296 <-explorer.exe             41428 <-msedgewebview2.exe
41428 <-msedgewebview2.exe            45092 <-msedgewebview2.exe
41428 <-msedgewebview2.exe            41356 <-msedgewebview2.exe
41428 <-msedgewebview2.exe            44988 <-msedgewebview2.exe
41428 <-msedgewebview2.exe            30420 <-msedgewebview2.exe
41428 <-msedgewebview2.exe            36324 <-msedgewebview2.exe
```
... I found the root 6296 is the ``explorer.exe``, which is not related at all to my application. (my-app-dev.exe or wails.exe)

So your method is wrong I guess!

**Reply by stffabi** [reply]
Please try to use `Process Explorer` from the SysInternals Tools to view the process tree. What do you see there?

**Reply by stffabi** [reply]
Task manager is not always right for process tree showing.

**Reply by KiddoV** [reply]
I used this method to retrieve the process tree for my app parent ID, it still show only 1 child process - same as the Task Manager or the  ``cmd.exe`` showed:
```go
// If the "withRoot" argument is set to "true", include the root process.
func getProcessTree(pid int, tree *[]process.Process, withRoot bool) error {
	proc, err := process.NewProcess(int32(pid))
	if err != nil {
		return err
	}
	children, err := proc.Children()
	if err == process.ErrorNoChildren {
		return nil
	}
	if err != nil {
		return err
	}
	// Iterate for each child process in reverse order.
	for i := len(children) - 1; i >= 0; i-- {
		child := children[i]
		// Call self to collect descendants.
		err := getProcessTree(int(child.Pid), tree, false)
		if err != nil {
			return err
		}
		// Add the child after it's descendants.
		*tree = append(*tree, *child)
	}
	// Add the root process to the end.
	if withRoot {
		*tree = append(*tree, *proc)
	}
	return nil
}
``` 
```go
	pTree := []process.Process{}
	getProcessTree(parentAppPID, &pTree, true)
	// Print collected info.
	for _, proc := range pTree {
		cmdLine, err := proc.Cmdline()
		if err != nil {
			panic(err)
		}
		fmt.Printf("%v\t| %v\r\n", proc.Pid, cmdLine)
	}
```
out put only show the wails.exe and my-app-dev.exe pid

**Reply by KiddoV** [reply]
> Please try to use `Process Explorer` from the SysInternals Tools to view the process tree. What do you see there?

I don't have that tool on my PC, can you provide the link to download it?

**Reply by stffabi** [reply]
> > Please try to use `Process Explorer` from the SysInternals Tools to view the process tree. What do you see there?
> 
> I don't have that tool on my PC, can you provide the link to download it?

https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer

Have you tried just starting the production build without `wails dev`? I currently wasn't able to see your behaviour, at least not with a production build and on Windows 11.

**Reply by KiddoV** [reply]
> Have you tried just starting the production build without wails dev?

You meant to build and run on exe file? I thought ``wails dev`` will build and run on ``my-app-dev.exe`` anyway? Is it not the same thing?

**Reply by stffabi** [reply]
> > Have you tried just starting the production build without wails dev?
> 
> You meant to build and run on exe file? I thought `wails dev` will build and run on `my-app-dev.exe` anyway? Is it not the same thing?

Yeah build and run, maybe there's some difference. I don't know...

**Reply by KiddoV** [reply]
@stffabi Yes, you are right, it groups all the processes within ``.exe`` file, not in the ``dev mode``.

Thank you!
